### PR TITLE
[Bug] R1 Accuracy: Fix `routed_scaling_factor` Double Mul Issue

### PR DESCRIPTION
**Pull request by yewentao256**
## Purpose

Fixes https://github.com/vllm-project/vllm/issues/24118 that is introduced by https://github.com/vllm-project/vllm/pull/23123

## Test

`lm_eval --model local-completions --model_args "base_url=http://127.0.0.1:9256/v1/completions,model=deepseek-ai/DeepSeek-R1,num_concurrent=256" --tasks gsm8k`

Origin

```bash
|Tasks|Version|     Filter     |n-shot|  Metric   |   |Value |   |Stderr|
|-----|------:|----------------|-----:|-----------|---|-----:|---|-----:|
|gsm8k|      3|flexible-extract|     5|exact_match|↑  |0.6626|±  |0.0130|
|     |       |strict-match    |     5|exact_match|↑  |0.5201|±  |0.0138|
```

Now
```bash
|Tasks|Version|     Filter     |n-shot|  Metric   |   |Value |   |Stderr|
|-----|------:|----------------|-----:|-----------|---|-----:|---|-----:|
|gsm8k|      3|flexible-extract|     5|exact_match|↑  |0.9530|±  |0.0058|
|     |       |strict-match    |     5|exact_match|↑  |0.9522|±  |0.0059|
```

### CODE REVIEW

File: vllm/model_executor/models/deepseek_v2.py [hunk 1]
```diff
@@ -186,15 +186,8 @@ def forward(self, hidden_states: torch.Tensor) -> torch.Tensor:
...
-            final_hidden_states = self.experts(
-                hidden_states=hidden_states,
-                router_logits=router_logits) * self.routed_scaling_factor
-        else:
-            # Fix FP16 overflow
-            # See DeepseekV2DecoderLayer for more details.
-            final_hidden_states = self.experts(hidden_states=hidden_states,
-                                               router_logits=router_logits)
+        final_hidden_states = self.experts(hidden_states=hidden_states,
+                                           router_logits=router_logits)
```
Comment (tlrmchlsmth) [inline vllm/model_executor/models/deepseek_v2.py]: This is handled here right?
https://github.com/vllm-project/vllm/blob/98aee612aa13155badc2747bd51b378d6e515958/vllm/model_executor/layers/fused_moe/fused_moe.py#L1014-L1015

Why does main not apply the `routed_scaling_factor` in the fp16 case? Is there an accuracy issue we need to handle?

Comment (shinoss) [reply vllm/model_executor/models/deepseek_v2.py]: looks like it's to prevent fp16 overflow issue for DeepSeek V2: https://github.com/vllm-project/vllm/pull/13232

by multiplying by routed scaling factor in all cases, https://github.com/vllm-project/vllm/pull/23123 might have introduced an accuracy regression for V2 as well, we should check this.

Comment (yewentao256) [reply vllm/model_executor/models/deepseek_v2.py]: Yes, so I changed another way to fix since deepseek v2 has a lot of logic for `routed_scaling_factor` outside

[the remark this answers is not in the document]
Comment (yewentao256) [reply vllm/model_executor/models/deepseek_v2.py]: Nice bot, fixed

**Review by shinoss — APPROVED** [review]
thanks for the fix

I think we also need this fix for dots1 and glm4_moe? both use `use_grouped_topk=True` so they should be affected by the double mul issue as well

https://github.com/vllm-project/vllm/blob/98aee612aa13155badc2747bd51b378d6e515958/vllm/model_executor/models/dots1.py#L163

https://github.com/vllm-project/vllm/blob/98aee612aa13155badc2747bd51b378d6e515958/vllm/model_executor/models/glm4_moe.py#L189

**Comment by yewentao256** [comment]
> thanks for the fix
> 
> I think we also need this fix for dots1 and glm4_moe? both use `use_grouped_topk=True` so they should be affected by the double mul issue as well
> 
> https://github.com/vllm-project/vllm/blob/98aee612aa13155badc2747bd51b378d6e515958/vllm/model_executor/models/dots1.py#L163
> 
> https://github.com/vllm-project/vllm/blob/98aee612aa13155badc2747bd51b378d6e515958/vllm/model_executor/models/glm4_moe.py#L189

@sarckk Nice find! Feel free to have a pr towards this.

**Review by houseroad — APPROVED** [review]
Thanks for the fix.

**Comment by WoosukKwon** [comment]
@yewentao256 Thanks for the fix. BTW, can you please add a test in our CI? I think we've seen similar accuracy issues 5+ times on DeepSeek...

**Comment by houseroad** [comment]
Besiding running full deepseek model, is it possible to run a smaller model to detect such issue, like DeepSeek v2, etc.
